File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/D-Безумные_монстры/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    a: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    d: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    left: list[int] = list(range(-1, n-1))
    right: list[int] = list(range(1, n+1))
    right[-1] = -1
    alive: list[bool] = [True] * n
    current_candidates: set[int] = set(range(n))
    output: list[int] = [0] * n
    for round_num in range(n):
        if not current_candidates: break
        to_die: list[int] = []
        next_candidates: set[int] = set()
        
        for j in current_candidates:
            if not alive[j]: continue
            L: int = left[j]
            while L != -1 and not alive[L]: L = left[L]
            R: int = right[j]
            while R != -1 and not alive[R]: R = right[R]
            
            damage: int = 0
            if L != -1: damage += a[L]
            if R != -1: damage += a[R]
            if damage > d[j]: to_die.append(j)
        
        for j in to_die:
            alive[j] = False
            original_L: int = left[j]
            original_R: int = right[j]
            if original_L != -1:
                right[original_L] = original_R
                if alive[original_L]: next_candidates.add(original_L)
            
            if original_R != -1:
                left[original_R] = original_L
                if alive[original_R]: next_candidates.add(original_R)
        output[round_num] = len(to_die)
        current_candidates = next_candidates
    sys.stdout.write('{}\n'.format(' '.join(map(str, output))))

if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/D-Безумные_монстры/solution.cpp
#include <iostream>
#include <vector>
#include <set>
#include <unordered_set>
#include <algorithm>

using namespace std;


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    int n;
    cin >> n;

    vector<int> a(n);
    for (int &x : a) cin >> x;

    vector<int> d(n);
    for (int &x : d) cin >> x;

    set<int> alive;
    for (int i = 0; i < n; ++i) alive.insert(i);

    vector<int> result(n, 0);
    unordered_set<int> current_candidates;
    for (int i = 0; i < n; ++i) current_candidates.insert(i);

    for (int round = 0; round < n; ++round) {
        if (current_candidates.empty()) break;
        unordered_set<int> to_die;
        unordered_set<int> next_candidates;
        for (int j : current_candidates) {
            if (alive.find(j) == alive.end()) continue;

            auto it = alive.find(j);
            int left_alive = -1, right_alive = -1;

            if (it != alive.begin()) {
                auto left_it = prev(it);
                left_alive = *left_it;
            }
            if (next(it) != alive.end()) {
                auto right_it = next(it);
                right_alive = *right_it;
            }

            int damage = 0;
            if (left_alive != -1) damage += a[left_alive];
            if (right_alive != -1) damage += a[right_alive];

            if (damage > d[j]) to_die.insert(j);
        }

        for (int j : to_die) {
            alive.erase(j);
            result[round]++;
            auto it = alive.lower_bound(j);
            if (it != alive.begin()) next_candidates.insert(*prev(it));
            if (it != alive.end()) next_candidates.insert(*it);
        }
        current_candidates = next_candidates;
    }

    for (int i = 0; i < n; ++i) cout << result[i] << ' ';
    cout << '\n';
}


int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    solution();
}